### ISSUE TITLE
Should a cascade after an expression treated as an explicit `call` tearoff be an error?

### ISSUE THREAD

**Issue by srawlins** [dart-lang/language#1873]
Take the following examples:

```dart
class C {
  void m() => print('m is called!');
  int call() => 7;
}

void main() {
  var c1 = C(); // instance of C
  int Function() c2 = C(); // tearoff of call method of instance of C
  int Function() c3 = C()..m(); // should be error? OK?
}
```

Currently neither the analyzer nor the CFE issues an error, and dart2js will in fact print "m is called!"

@leafpetersen noted in a conversation that the `c3` case should result in an error because the context type (`int Function()`) changes the meaning of `C()`.

**Comment by eernstg** [comment]
Given that the context type for `C()` in both cases is `int Function()`, we should prepend a `.call` to the cascade receiver, and then `m()` should be a compile-time error.

You could say that the effect is spooky, but it is just another situation where the context type changes the meaning of an expression: `List<int> xs = []..length` changes the list to be a `List<int>` at run time rather than a `List<dynamic>`, so we already have the effect that context types matter (statically and dynamically) in lots of other ways.

We could of course make an exception, based on the opinion that it is too spooky to change `C()` to a function.

**Comment by lrhn** [comment]
Notice that what @eernstg states here is not how our compilers currently work.
They instead wait for the value of the cascade to perform the implicit `.call` tear-off.

That allows for code like this:
```dart
class MyCheapMockClass {
  int returnValue = 0;
  int call() => returnValue;
}
...
int Function() f = MyCheapMockClass()..returnValue = 42;
```
where you use a cascade to initialize the callable object before it gets implicitly torn off.

**Comment by eernstg** [comment]
Right, but we do not have the notion of a "direct context type" and a "context type which was propagated from an expression to one of its subexpressions", and in order to make the specification describe the current behavior we would need that plus some extra rules about _which_ propagations should be allowed to cause a `.call` insertion (presumably, we wouldn't want to eliminate `.call` insertion for collection literal elements or parentheses, and I'm not sure about conditional expressions).

**Comment by natebosch** [comment]
If `int Function() c3 = C()..m();` is an error, what is the correct way to write it? Is it to make the tearoff of `call` explicit with `int Function() c3 = (C()..m()).call;`?

**Comment by lrhn** [comment]
Making the `.call` explicit would put the receiver expression into a position where it has no context type.
(Just hope that that context type isn't needed for anything else. It *probably* isn't.)

**Comment by eernstg** [comment]
I think it's ok to rewrite as `(C()..m()).call`: If we accept the premise that the context type for the entire expression is `Function` or a function type, and the static type of the receiver is not a function type, then we'll never be in the situation where that context type is used for anything relative to the receiver.

I actually like the fact that `(C()..m()).call` has an _explicit_ `.call`, which could be taken as a point in favor of keeping things as specified: It's good that `C()..m()` with context type `Function` or a function type is generally a compile-time error (because that function doesn't have an `m` member), and then the developer will have to add `().call` around the whole cascade, or use something like this:

```dart
var c2dot9 = C()..m(); // Just a normal cascade.
int Function() c3 = c2dot9; // Adds `.call`, but doesn't involve further magic.
```

In other words, we _could_ introduce new rules about different kinds of context types, and we _could_ then turn `int Function() c3 = C()..m();` into something that "just works", but I'm not convinced that this will make us happier. ;-)

**Comment by srawlins** [comment]
Quick note: there is code in the material components which depends on the current behavior, that the cascade is performed on the interface type, and the resulting expression, the cascade expression, is still treated as the call tear-off: https://github.com/angulardart/angular_components/blob/master/examples/material_input_example/lib/material_input_demo.dart#L98

Unless the explicit parentheses around the cascade are what guarantees that behavior.

**Comment by lrhn** [comment]
Can anyone say precisely how the implementations work? What rules do they use to determine whether to add the `.call`, which makes it end up as `Function f = (C()..member()).call;`?

**Comment by eernstg** [comment]
> would have to add `(...).call` around the cascade ... it'd be breaking

Right, but I took into account that it is likely to be a small amount of breakage. So that made me hope that we could allow it.

It may be an anomaly that the cascade allows us to use an expression two or more times, but it is also the main purpose of the cascade construct in the first place.

The language specification does not yet specify how context types propagate to subexpressions. There is no definition of 'context type' in other specification documents, except for (1) special cases for pattern related constructs in `patterns-feature-specification.md`, (2) special cases for numeric expressions in `number-operation-typing.md`, (3) special cases for `e` in `return e` in function literals, in `inference.md` (plus in some commented-out parts of `inference.md` near the end). So we don't have a comprehensive specification of context types.

In any case, we probably don't want to have a rule where the context type does _not_ propagate from the cascade to the cascade receiver. The example I've used for this is `List<int> xs = []..add(1);` where we do not want `[]` to be inferred as `<dynamic>[]`.

We _do_ actually specify cascades in terms of a `let` construct. But when we fill in the missing parts of a specification of context types, I would expect the context type of the cascade as a whole to propagate to the receiver.

**Comment by eernstg** [comment]
Note that `.call` insertion on the individual branches of a conditional came up in https://github.com/dart-lang/sdk/issues/47787. Here is a minimal example that shows the failure:

```dart
class A {
  void call() {}
}

bool b = true;

void main() {
  // The error is eliminated if we change `A()` to `A().call`.
  Function f = b ? (){} : A(); // "A value of type 'Object' can't be assigned to a variable of type 'Function'."
}
```
